File: protocol.py
from dataclasses import dataclass, field
from enum import Enum
import json
from typing import Any
from settings import PROTOCOL_MAGIC, PROTOCOL_VERSION


class MessageType(Enum):

    FILE_OFFER = "FILE_OFFER"
    FILE_ACCEPT = "FILE_ACCEPT"
    FILE_REJECT = "FILE_REJECT"


@dataclass
class Message:
    message_type: MessageType
    payload: dict[str, Any] = field(default_factory=dict) # | None = None

def encode_message(message: Message) -> bytes:
    # Serialize a Message object into UTF-8 encoded JSON bytes.
    data = {
        "magic": PROTOCOL_MAGIC,
        "version": PROTOCOL_VERSION,
        "type": message.message_type.value,
        "payload": message.payload or {},
    }

    return json.dumps(data).encode("utf-8")

def decode_message(raw_data: bytes) -> Message:
    # Deserialize UTF-8 encoded JSON bytes into a Message object.
    data = json.loads(raw_data.decode("utf-8"))

    if data.get("magic") != PROTOCOL_MAGIC:
        raise ValueError("Invalid protocol identifier.")

    if data.get("version") != PROTOCOL_VERSION:
        raise ValueError(f"Invalid protocol version: {data.get('version')}")

    return Message(
        message_type=MessageType(data["type"]),
        payload=data["payload"],
    )

File: android/app/src/main/java/com/lanfiletransfer/android/Protocol.kt
package com.lanfiletransfer.android

import org.json.JSONObject
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.InputStream
import java.io.OutputStream

/**
 * Mirrors LANFileTransfer's protocol.py.
 *
 * IMPORTANT: PROTOCOL.md (the repo's own docs) describes the JSON key as
 * "message_type", but the actual Python implementation (protocol.py) sends
 * the key as "type". This file matches the real code, not the docs.
 */
object ProtocolConstants {
    const val MAGIC = "LANFILETRANSFER"
    const val VERSION = 1
}

data class Message(val type: String, val payload: JSONObject)

object MessageCodec {

    fun encode(message: Message): ByteArray {
        val obj = JSONObject()
        obj.put("magic", ProtocolConstants.MAGIC)
        obj.put("version", ProtocolConstants.VERSION)
        obj.put("type", message.type)
        obj.put("payload", message.payload)
        return obj.toString().toByteArray(Charsets.UTF_8)
    }

    fun decode(data: ByteArray): Message {
        val obj = JSONObject(String(data, Charsets.UTF_8))
        if (obj.optString("magic") != ProtocolConstants.MAGIC) {
            throw IllegalArgumentException("Invalid protocol identifier.")
        }
        if (obj.optInt("version") != ProtocolConstants.VERSION) {
            throw IllegalArgumentException("Invalid protocol version: ${obj.opt("version")}")
        }
        val type = obj.getString("type")
        val payload = obj.optJSONObject("payload") ?: JSONObject()
        return Message(type, payload)
    }
}

/**
 * Every packet is a 4-byte big-endian ("network byte order") length prefix
 * followed by that many bytes of UTF-8 JSON. This matches struct.pack("!I", ...)
 * on the Python side. File bytes themselves are NOT wrapped in this framing;
 * they are streamed raw right after a FILE_ACCEPT.
 */
object PacketIO {

    fun sendPacket(out: OutputStream, message: Message) {
        val data = MessageCodec.encode(message)
        val dos = DataOutputStream(out)
        dos.writeInt(data.size) // big-endian, matches "!I"
        dos.write(data)
        dos.flush()
    }

    /** Returns null on clean connection close (mirrors recv_exact returning None). */
    fun receivePacket(input: InputStream): Message? {
        val dis = DataInputStream(input)
        val length = try {
            dis.readInt()
        } catch (e: Exception) {
            return null
        }
        if (length < 0 || length > 64 * 1024 * 1024) {
            // Sanity cap for a JSON control message (not the file stream itself).
            throw IllegalArgumentException("Unreasonable message length: $length")
        }
        val buffer = ByteArray(length)
        dis.readFully(buffer)
        return MessageCodec.decode(buffer)
    }
}
